packages/swing-store/src/snapStore.js

```diff
@@ -356,19 +407,130 @@ export function makeSnapStore(
     sqlDeleteSnapshotByHash.run(vatID, hash);
   }
 
-  const sqlDumpActiveSnapshots = db.prepare(`
+  const sqlGetSnapshotMetadata = db.prepare(`
+    SELECT vatID, endPos, hash, uncompressedSize, compressedSize
+    FROM snapshots
+    WHERE inUse = ?
+    ORDER BY vatID, endPos
+  `);
+
+  /**
+   * @param {boolean} includeHistorical
+   * @yields {[key: string, value: string]}
+   * @returns {Iterable<[key: string, value: string]>}
+   */
+  function* getExportRecords(includeHistorical) {
+    for (const rec of sqlGetSnapshotMetadata.iterate(1)) {
+      const exportRec = snapshotRec(
+        rec.vatID,
+        rec.endPos,
+        rec.hash,
+        rec.uncompressedSize,
+      );
+      const exportKey = snapshotMetadataKey(rec);
+      yield [exportKey, JSON.stringify(exportRec)];
+      yield [currentSnapshotMetadataKey(rec), snapshotArtifactName(rec)];
+    }
+    if (includeHistorical) {
+      for (const rec of sqlGetSnapshotMetadata.iterate(0)) {
+        const exportRec = snapshotRec(
+          rec.vatID,
+          rec.endPos,
+          rec.hash,
+          rec.uncompressedSize,
+        );
+        yield [snapshotMetadataKey(rec), JSON.stringify(exportRec)];
+      }
+    }
+  }
+
+  function* getArtifactNames(includeHistorical) {
+    for (const rec of sqlGetSnapshotMetadata.iterate(1)) {
+      yield snapshotArtifactName(rec);
+    }
+    if (includeHistorical) {
+      for (const rec of sqlGetSnapshotMetadata.iterate(0)) {
+        yield snapshotArtifactName(rec);
+      }
+    }
+  }
+
+  /**
+   * @param {string} name  Artifact name of the snapshot
+   * @param {SwingStoreExporter} exporter  Whence to get the bits
+   * @param {object} info  Metadata describing the artifact
+   * @returns {Promise<void>}
+   */
+  async function importSnapshot(name, exporter, info) {
+    const parts = name.split('.');
+    const [type, vatID, rawEndPos] = parts;
+    assert(
+      parts.length === 3 && type === 'snapshot',
+      `expected snapshot name of the form 'snapshot.{vatID}.{endPos}', saw '${name}'`,
+    );
+    assert(
+      info.vatID === vatID,
+      `snapshot name says vatID ${vatID}, metadata says ${info.vatID}`,
+    );
+    const endPos = Number(rawEndPos);
+    assert(
+      info.endPos === endPos,
+      `snapshot name says endPos ${endPos}, metadata says ${info.endPos}`,
+    );
+
+    const artifactChunks = exporter.getArtifact(name);
```

Comment:
This still feels like it's sitting at the wrong boundary. The same artifact name validation logic is duplicated in swingStore's `getArtifact`, and this is exposing the `exporter` concept to what I consider the internal `snapStore`?

Why not move name & metadata validation logic in the swingStore, and change this to simply be taking the chunks and parsed (and pre-validated) info?

```suggestion
  /**
   * @param {AsyncIterable<Uint8Array>} artifactChunks
   * @param {{ vatID, endPos, hash, size }} info  Metadata describing the artifact
   * @returns {Promise<void>}
   */
  async function importSnapshot(artifactChunks, info) {
    const {vatID, endPos} = info;
```

Reply:
The idea is that the validation of snapshots and snapshot metadata is the responsibility of the snapStore, while validation of transcripts and transcript metadata is the responsibility of the transcriptStore.  If we move that logic into the swingStore, then it becomes more brittle under maintenance, especially as we contemplate having more kinds of artifacts in the future (e.g., bundles), each of which may be handled by its own sub-store implementation and have its own metadata.